api/src/main/java/jakarta/servlet/ServletContext.java

```diff
@@ -1436,4 +1523,97 @@
      * @since Servlet 4.0
      */
     public void setResponseCharacterEncoding(String encoding);
+
+    /**
+     * The mode in which this context handles the encoding and decoding of URI paths. This mode affects the behavior of:
+     * <ul>
+     * <li>{@link HttpServletRequest#getContextPath()}</li>
+     * <li>{@link HttpServletRequest#getServletPath()}</li>
+     * <li>{@link HttpServletRequest#getPathInfo()}</li>
+     * <li>{@link HttpServletRequest#getPath()}</li>
+     * <li>{@link HttpServletRequest#getRealPath(String)}</li>
+     * <li>{@link HttpServletRequest#getRequestDispatcher(String)}</li>
+     * <li>{@link ServletContext#getContextPath()}</li>
+     * <li>{@link ServletContext#getContext(String)}</li>
+     * <li>{@link ServletContext#getRequestDispatcher(String)}</li>
+     * <li>{@link ServletContext#getRealPath(String)}</li>
+     * <li>{@link ServletContext#getPath(String)}</li>
+     * <li>{@link ServletContext#getResource(String)}</li>
+     * <li>{@link ServletContext#getResourceAsStream(String)}</li>
+     * <li>{@link ServletContext#getResourceAsPath(String)}</li>
+     * <li>{@link ServletContext#getResourcePaths(String)}</li>
+     * <li>{@link AsyncContext#dispatch(String)}</li>
+     * <li>{@link AsyncContext#dispatch(ServletContext, String)}</li>
+     * </ul>
+     * for the purposes of these modes:
+     * <ul>
+     * <li>a 'reserved character' is a character defined as <code>reserved</code> by
+     * <a href="https://datatracker.ietf.org/doc/html/rfc3986#section-2.2">RFC3986 Section 2.2</a> or the ';'
+     * character.</li>
+     * <li>a 'path parameter' is the part of a URI segment as defined by the obsoleted
+     * <a href="https://datatracker.ietf.org/doc/html/rfc2396#section-3.3">RFC2396 Section 3.3</a>.</li>
+     * </ul>
+     * 
+     * @since Servlet 6.0
+     */
+    public enum PathCodingMode {
+        /**
+         * Effected methods will return/accept paths as was done prior to the introduction of the PathCodingMode in Servlet 6.0.
+         * The path returned by {@link HttpServletRequest#getContextPath()} is not decoded. Paths returned by the methods
+         * {@link HttpServletRequest#getServletPath()} and {@link HttpServletRequest#getPathInfo()} are fully decoded and have
+         * all path parameters removed. Decoding may include ambiguous characters such as a decoded <code>%2F</code>, which is
```

Comment:
I don't see how we can avoid decoding %2F as the current api says that the strings are decoded, so any container that doesn't decode is in violation.   For jetty we by default reject such requests with a 400 unless the application is configured to say that it is prepared to accept such  decoded strings.

For me that is the key difference between `LEGACY` and `DECODED` mode: the former will allow containers to act as they have previous, whilst the later is a well defined mode that never returns a decoded %2F because they have been rejected with a 400.

Reply:
By default Undertow does not decode these, however if I had my time again I would respond with a 400 instead. Either way we will need a container specific option to allow this, as some people want it, but IMHO we should be strict about rejecting this in the spec.

Although something I just thought of is that we likely need to be clearer about reserved characters in the different parts of the URI, e.g. %2F is likely fine as part of a query string. A query string does not really have the same security issues as the actual URI, so rejecting encoded reserved characters in the query string is almost certainly the wrong thing to do.